main/src/ca/uwaterloo/flix/language/phase/unification/Unification.scala

```diff
@@ -463,6 +474,27 @@ object Unification {
       Ok((s, econstrs, renv, lenv, purifiedEff))
     }
 
+  /**
+    * Purifies the given tvar in the type and unbinds variables in the substitution that reside in a level
+    * that is deeper than the current level.
+    */
+  def purifyEffAndUnbind(tvar: Type.Var, eff: Type)(implicit level: Level, flix: Flix): InferMonad[Type] = {
+    InferMonad { case (s, econstrs, renv, lenv) =>
+
+      var s2 = s
+
+      val res = s(eff).map {
+        // Case 1: The region. Purify it.
+        case tv if tv.sym == tvar.sym => Type.Pure
+        // Case 2: A variable from inside the region. Unbind it.
```

Comment:
I still worry this would leave variables free inside the region?

Reply:
They will only be free if they would be free anyway. Consider:

```scala
def any(): Unit \ a = ???

def foo(): Unit \ IO =
    region r {
        any()        // Unit \ ef   (where ef is free)
    }                // Unit \ ef

Here ef will unify with IO and no longer be free.